draft-irtf-cfrg-hash-to-curve.md

```diff
@@ -356,869 +524,832 @@ document are to be interpreted as described in {{RFC2119}}.
+not sufficient to get uniformity, but it can be used as a building block as
+described in {{rom}}.
 
-- CMOV(a, b, c): If c = 1, return a, else return b.
+# Utility Functions {#utility}
 
-  Common software implementations of constant-time selects assume c = 1 or c = 0. CMOV
-  may be implemented by computing the desired selector (0 or 1) by ORing all bits of c
-  together. The end result will be either 0 if all bits of c are zero, or 1 if at least
-  one bit of c is 1.
+Algorithms in this document make use of utility functions described below.
 
-- CTEQ(a, b): Returns a == b.
-  Inputs a and b must be the same length (as bytestrings) and the comparison
-  must be implemented in constant time.
+- is_square(x, q): It returns True whenever the value x is a quadratic residue
+  in a field of order q. Due to Euler's criterion, this function can be
+  calculated in constant time as
+
+  ~~~
+    is_square(x, q) := { True,  if x^((q-1)/2) is 0 or 1;
+                       { False, otherwise.
+  ~~~
+
+  For clarity, it is generally preferred to use this formula directly, and
+  annotate its usage with this definition.
+
+- sqrt(x, q): The sqrt operation is a multi-valued function, i.e. there exist
+  two roots of x whenever x is a quadratic residue. To maintain compatibility
+  across implementations, a single-valued sqrt function is necessary. One way
+  to get such a function is by distinguishing a principal square root through a
+  predicate that only one of the roots holds.
+  Alternatively, an implementation of sqrt can use fixed formulas for its
+  calculation. For instance, in prime fields, the square root of a quadratic
+  residue x can be obtained as follows
+  - If q=3 (mod 4), sqrt(x, q) := x^((q+1)/4).
+  - If q=5 (mod 8), set z := x^((q+3)/8) and verify that z^2 = -x, if so, update
+    z = z\*sqrt(-1). Finally, sqrt(x, q) := z.
+
+  For extension fields, there exist methods that can be used in replacement,
+  see {{Adj13}}, {{SC85}}. Regardless the method chosen, the sqrt function
+  should be performed in constant time.
+
+- CMOV(a, b, c): If c=0, CMOV returns a, otherwise returns b. To prevent against
+  timing attacks, this operation must run in constant time without revealing the
+  value of c. Commonly some implementations assume the selector be c=1 or c=0;
+  thus, given a bitstring C, the desired selector c can be computed by OR-ing
+  all bits of C together. The resulting selector will be either 0 if all bits
+  of C are zero, or 1 if at least one bit of C is 1.
+
+- I2OSP and OS2IP: These functions are used to convert an octet string to
+  and from a non-negative integer {{RFC8017}}.
+
+- a \|\| b: denotes the concatenation of bitstrings a and b.
+
+# Hashing to a Finite Field {#hashtobase}
+
+The hash2base(x) function maps an arbitrary-large bitstring x into an element
+of a field F. Hence, it is parametrized by F and H, where H is a
+cryptographic hash function which outputs at least floor(log2(p)) + 1 bits.
```

Comment:
Based on discussion, shouldn't this be at least `floor(log2(p))+64`?

Reply:
If so, can we also move the rationale for this requirement to an appendix, perhaps with a sketch from Michele's proof?

Reply:
This will be resolved in another PR. Currently #99 will track this issue.